solidity/random-beacon/contracts/libraries/Relay.sol

```diff
@@ -160,12 +200,25 @@ library Relay {
             "Invalid entry"
         );
 
-        // TODO: Use submitterIndex, firstEligibleIndex and lastEligibleIndex
-        //       to prepare an array of addresses which should be kicked from
-        //       the sortition pool for 2 weeks.
-
-        // TODO: If soft timeout has elapsed, take bleeding into account
-        //       and slash all members appropriately.
+        // Get the list of members addresses which should be punished due to
+        // not submitting the entry on their turn.
+        address[] memory punishedMembers = getPunishedMembers(
+            self,
+            submitterIndex,
+            firstEligibleIndex,
+            group,
+            groupSize
+        );
+        self.sortitionPool.removeOperators(punishedMembers);
+
+        // If the soft timeout has been exceeded apply stake slashing for
+        // all group members. Note that `getSlashingFactor` returns the
+        // factor multiplied by 1e18 to avoid precision loss. In that case
+        // the final result needs to be divided by 1e18.
+        uint256 slashingAmount = (getSlashingFactor(self, groupSize) *
+            self.relayEntrySubmissionFailureSlashingAmount) / 1e18;
+        // slither-disable-next-line reentrancy-events
+        self.staking.slash(slashingAmount, group.members);
```

Comment:
If we decide to slash only in `RandomBeacon` (which IMO would be a nice pattern), we could return slashing amount from this function and call `staking.slash` in `RandomBeacon` only if this amount is non-zero.

Reply:
Sure. I will handle it in follow-up https://github.com/keep-network/keep-core/pull/2671 where I do the same for `inactiveMembers`. This way we don't block here with tests and we don't introduce new conflicts.